docs/code/client_middleware_cookbook.py

```diff
@@ -0,0 +1,132 @@
+"""This is a collection of semi-complete examples that get included into the cookbook page."""
+
+import asyncio
+import logging
+import time
+from collections.abc import AsyncIterator, Sequence
+from contextlib import asynccontextmanager, suppress
+
+from aiohttp import (
+    ClientHandlerType,
+    ClientRequest,
+    ClientResponse,
+    ClientSession,
+    TCPConnector,
+)
+from aiohttp.abc import ResolveResult
+from aiohttp.tracing import Trace
+
+
+class SSRFError(Exception):
+    """A request was made to a blacklisted host."""
+
+
+async def retry_middleware(
+    req: ClientRequest, handler: ClientHandlerType
+) -> ClientResponse:
+    for _ in range(3):  # Try upto 3 times
+        resp = await handler(req)
+        if resp.ok:
+            return resp
+    assert False  # Unreachable
```

Comment:
Should this raise a `ClientError`?

Reply:
Could do, it really makes no difference. It's just that mypy can't tell that the loop will always execute. A future release could change that one day.

Reply:
bace3a9a9